### PR TITLE
Nice path decomposition was made

### PR DESCRIPTION
**Pull request by bazillo**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -16,5 +16,12 @@ add_library(centroid_tree src/centroid-tree.cpp)
 
 add_executable(test src/test.cpp)
 add_test(NAME test0 COMMAND test)
+
+target_link_libraries(test PUBLIC path_dec)
+add_library(nice_path_dec src/nice-path-decomposition/nice-path-decomposition.cpp)
+add_executable(nice_test src/nice-path-decomposition/test.cpp)
```
Comment (tunyash) [inline CMakeLists.txt]: Let's put the class and the test file in the root of src.

File: src/nice-path-decomposition/test.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+            {2, 5},
+            {2, 6}
+    };
+
+    Graph g(edges.begin(), edges.end(), n);
+    CentroidTree c(g);
+    NicePathDecomposition p(c.get_path_decomposition(g));
+
+    auto a = p.get_bags();
+    auto nice_a = p.get_nice_bags();
```
Comment (tunyash) [inline src/nice-path-decomposition/test.cpp]: Does it checks the correctness? Seems like the correctness of the nice representation is never checked. Please add a method for it.

File: src/nice-path-decomposition/nice-path-decomposition.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    nice_transform();
+}
+
+
+NicePathDecomposition::NicePathDecomposition(PathDecomposition pathDecomposition) : PathDecomposition(
+        pathDecomposition) {
+    nice_transform();
+}
+
+void NicePathDecomposition::nice_transform() {
```
Comment (tunyash) [inline src/nice-path-decomposition/nice-path-decomposition.cpp]: Does it change the initial _bags field? The word 'transform' suggest so.

File: src/nice-path-decomposition/nice-path-decomposition.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        edge_t edge;
+
+        Bag(bag_types bag_type, vertex_t v) : type(bag_type), v(v) {}
+
+        Bag(bag_types bag_type, edge_t edge) : type(bag_type), edge(edge) {}
+    };
+
+    std::vector<Bag> get_nice_bags();
+
+    NicePathDecomposition(std::vector<std::vector<vertex_t>> &bags, Graph &g);
```
Comment (tunyash) [inline src/nice-path-decomposition/nice-path-decomposition.h]: Add docstrings for constructors and public methods. Make sure you mention that the bags here are not expected to be nice.

File: src/nice-path-decomposition/nice-path-decomposition.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#include "../centroid-tree.h"
+
+
+class NicePathDecomposition : public PathDecomposition {
```
Comment (tunyash) [inline src/nice-path-decomposition/nice-path-decomposition.h]: Add a docstring: a definition of a nice path decomposition and the purpose of this class.

File: src/nice-path-decomposition/test.cpp [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#include "nice-path-decomposition.h"
+//#include "../centroid-tree.h"
```
Comment (tunyash) [inline src/nice-path-decomposition/test.cpp]: Remove

File: src/nice-path-decomposition/nice-path-decomposition.cpp [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    std::vector<vertex_t> prev_bag = {};
+    std::set<vertex_t> prev_set = {};
+    for (const std::vector<vertex_t>& bag : _bags) {
+        std::set<vertex_t> vertices;
+        for (vertex_t v : bag) vertices.insert(v);
+        for (int i = 0; i < prev_bag.size(); i++) {
```
Comment (tunyash) [inline src/nice-path-decomposition/nice-path-decomposition.cpp]: This is a for loop over a changing vector that manipulates i inside the loop. This makes code harder to read and more bug-prone than necessary. Here you compute the difference between the set of nodes in the current bag and in the next one. There is a function in the STD algorithm library that does that.

File: src/nice-path-decomposition/nice-path-decomposition.cpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+            if (vertices.count(v)) vertices.erase(v);
+            else {
+                prev_bag.erase(prev_bag.begin() + i);
+                prev_set.erase(v);
+                i--;
+                _nice_bags.emplace_back(bag_types::REMOVE_VERTEX, v);
+            }
+
+        }
+        for (vertex_t v : vertices) {
```
Comment (tunyash) [inline src/nice-path-decomposition/nice-path-decomposition.cpp]: Here you need to iterate over difference between current bag and the previous one which also can be done via the standard function

**Review by tunyash — CHANGES_REQUESTED** [review]
Please add a niceness check method and address the comments (as well as the failing unit test check)

File: src/nice-path-decomposition.cpp [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    std::vector<bool> removed(n, false);
+    std::map<edge_t, bool> edge_used;
+    for (Bag bag : _nice_bags) {
+        if (bag.type == ADD_VERTEX) added[bag.v] = true;
+        else if (bag.type == REMOVE_VERTEX) removed[bag.v] = true;
+        else edge_used[bag.edge] = true;
+    }
+    for (vertex_t v = 0; v < n; v++) {
+        if (!added[v]) {
+            std::cerr << "vertex " << v << " haven't been added";
```
Comment (tunyash) [inline src/nice-path-decomposition.cpp]: throw Exception instead (and in all the remaining places as well)

File: src/nice-path-decomposition.cpp [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+
+NicePathDecomposition::NicePathDecomposition(PathDecomposition pathDecomposition) : PathDecomposition(
+        pathDecomposition) {
+    fill_nice_bags();
+}
+
+
+bool NicePathDecomposition::Bag::operator<(Bag b) {
```
Comment (tunyash) [inline src/nice-path-decomposition.cpp:45]: I don't quite get what is it for, is it just to have some comparator or does it have a more sophisticated purpose? If it is just to be able to use sets, I suggest simply joining all fields into a tuple and comparing the tuples.

File: src/nice-path-decomposition.h [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        Bag() = default;
+
+        bool operator<(Bag b);
+        // Use it to sort bags
+    };
+
+
+    std::vector<Bag> get_nice_bags();
+
+    NicePathDecomposition(std::vector<std::vector<vertex_t>> &bags, Graph &g);
```
Comment (tunyash) [inline src/nice-path-decomposition.h:54]: should bags be const?

**Review by tunyash — APPROVED** [review]
LGTM, given the comments are addressed

**Comment by elproblema** [comment]
Is it possible to cast NicePathDecomposition to PathDecomposition or just get normal form of PathDecomposition?

**Comment by elproblema** [comment]
unit-tests.yml has to run successfully

File: CMakeLists.txt [hunk 12]
```diff
@@ -13,8 +13,16 @@ endif(UNIX)
...
 add_executable(test src/test.cpp)
-add_test(NAME test0 COMMAND test)
+add_executable(nice_test src/nice-test.cpp)
+
 target_link_libraries(test PUBLIC centroid_tree)
-target_link_libraries(centroid_tree PUBLIC path_dec)
\ No newline at end of file
+target_link_libraries(disj_paths_bd_tw PUBLIC centroid_tree)
+target_link_libraries(centroid_tree PUBLIC path_dec)
+target_link_libraries(nice_test PUBLIC nice_path_dec path_dec centroid_tree)
```
Comment (elproblema) [inline CMakeLists.txt:25]: I suppose its better to add nice-test.cpp to test executable. As well for unit-tests.yml

File: .github/workflows/unit-tests.yml [hunk 13]
```diff
@@ -51,8 +51,12 @@ jobs:
...
           cmakeListsTxtPath: '${{ github.workspace }}/CMakeLists.txt'
           useVcpkgToolchainFile: true
-          buildWithCMakeArgs: '-- -v'
+          buildWithCMakeArgs: '--trace -- -v'
           buildDirectory: '${{ github.workspace }}'
       - name: Run unit-tests
-        run: ./test
+        run: |
+          ./nice_test 
+          ./test
```
Comment (elproblema) [inline .github/workflows/unit-tests.yml:59]: Delete ./nice_test

File: src/centroid-tree.cpp [hunk 14]
```diff
@@ -113,7 +113,8 @@ PathDecomposition CentroidTree::get_path_decomposition(Graph &g) {
 
 CentroidTree::CentroidTree(Graph &g) {
     Tree t;
-    treedec::exact_decomposition_cutset(g, t);
+    Graph temp = g;
+    treedec::exact_decomposition_cutset(temp, t);
```
Comment (elproblema) [inline src/centroid-tree.cpp:117]: Add a new issue about it

File: src/nice-path-decomposition.cpp [hunk 15]
```diff
@@ -0,0 +1,100 @@
+#include "nice-path-decomposition.h"
+#include "../tdlib/src/iter.hpp"
+
+std::vector<NicePathDecomposition::Bag> NicePathDecomposition::get_nice_bags() {
+    return _nice_bags;
```
Comment (elproblema) [inline src/nice-path-decomposition.cpp:5]: Should be const &

File: src/nice-path-decomposition.cpp [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    if (type == ADD_VERTEX) v1 = v;
+    else v1 = std::max(edge.m_target, edge.m_source);
+    if (b.type == ADD_VERTEX) v2 = b.v;
+    else v2 = std::max(b.edge.m_target, b.edge.m_source);
+    return (v1 < v2) || (v1 == v2 && type == ADD_VERTEX);
+}
+
+void NicePathDecomposition::fill_nice_bags() {
+    int n = _g.m_vertices.size();
+    std::vector<int> left(n, n + 1);
```
Comment (elproblema) [inline src/nice-path-decomposition.cpp:59]: Make variables names the same as for path-decomposition.h

File: src/nice-path-decomposition.h [hunk 17]
```diff
@@ -0,0 +1,47 @@
+#include "centroid-tree.h"
+
+
+class NicePathDecomposition : public PathDecomposition {
+    /*
+        NicePathDecomposition is a path decomposition but every bag has one of three types: ADD_VERTEX, REMOVE_VERTEX, ADD_EDGE.
+        The porpoise of using it is that it will help to calculate the dynamic.
```
Comment (elproblema) [inline src/nice-path-decomposition.h:7]: Lexical error in docstring

File: src/nice-path-decomposition.cpp [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+bool NicePathDecomposition::Bag::operator<(Bag b) {
+    if (type == REMOVE_VERTEX && b.type != REMOVE_VERTEX) return true;
+    if (type != REMOVE_VERTEX && b.type == REMOVE_VERTEX) return false;
+    if (type == REMOVE_VERTEX && b.type == REMOVE_VERTEX) return v < b.v;
+    int v1, v2;
+    if (type == ADD_VERTEX) v1 = v;
+    else v1 = std::max(edge.m_target, edge.m_source);
+    if (b.type == ADD_VERTEX) v2 = b.v;
+    else v2 = std::max(b.edge.m_target, b.edge.m_source);
+    return (v1 < v2) || (v1 == v2 && type == ADD_VERTEX);
```
Comment (elproblema) [inline src/nice-path-decomposition.cpp:54]: In your operator sometimes commutativity is allowed, which might cause errors

**Pull request merged**